### PR TITLE
Feature psi

### PR DESCRIPTION
**Pull request by powerlink-team-kalycito**
This pull requests contains the components for openSAFETY demo integration rebased on openPOWERLINK 2.5.0.

Following components are added
1) SocketWrapper module for IP Stack 
2) Upgrade demo-sn-gpio HW design to use openMAC 1.0.2

The changes are tested with openPOWERLINK 2.5.0 rc1.

Known Issues:
1. The Safety operation could not be verified because of a board issue in the safety application processor board at Kalycito.

### CODE REVIEW

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    if(!memcmp(&macAddr , &macAddrZero , sizeof(eth_addr)))
+    {
+         printf(" Inside socketwrapper_arpQuery\n");
+        // send ARP request when MAC is unknown!
+        ipArpRequest(remoteIpAddress_p, &macAddr);
+        ret = kErrorSdoUdpArpInProgress;
+    }
+       return ret;
+
+}*/
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: Why is this commented?

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c [hunk 2]
```diff
@@ -85,11 +93,27 @@ provided from the Virtual Ethernet driver NoOs.
...
 /**
- * \brief Instance structure of the wrapper module
+\brief  Receive buffer descriptor
+
+This structure defines the receive buffer descriptor.
+*/
+typedef struct
+{
+    BOOL            fIpStackOwner;      ///< TRUE if the IP stack owns the buffer
+                                        //<TODO: Use here same type like in \ref ip_packet_typ!
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: Is that really a TODO?!

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c [hunk 3]
```diff
@@ -115,123 +137,74 @@ static void edrv2veth_freePacket(ip_packet_typ *pPacket);
...
-    {
-        // return number of bytes sent!
-        plkLen = pPacket->length;
-
-        // call free function
-        pFctFree(pPacket);
-    }
-
-    return plkLen;
+	edrv2vethInstance_l.pIpStack = NULL;
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: Wrong indenting?!

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c [hunk 4]
```diff
@@ -265,74 +238,186 @@ static void edrv2veth_changeAddress(UINT32 ipAddr_p, UINT32 subNetMask_p, UINT16
...
+    }
 
-    // TODO: Address manipulation only works when used with openMAC!
-    //       (The length field needs to be before the packet!)
-    pPacket = GET_TYPE_BASE(ip_packet_typ, data, pFrame_p);
+    if (i == IP_RX_BUF_CNT)
+    {
+    // No free buffer found => ignore frame
+    PRINTF("%s (Err/Warn): No free buffer found, Frames ignored\n");
+    return kErrorOk;
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: indenting

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c [hunk 5]
```diff
@@ -265,74 +238,186 @@ static void edrv2veth_changeAddress(UINT32 ipAddr_p, UINT32 subNetMask_p, UINT16
...
     tOplkError     ret = kErrorOk;
     INT            rcvStatus;
     ip_packet_typ* pPacket;
+    INT            i;
 
-    UNUSED_PARAMETER(frameSize_p);
+    for (i=0; i<IP_RX_BUF_CNT; i++)
+    {
+        if (edrv2vethInstance_l.aRxBuffer[i].fIpStackOwner)
+        break;
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: indenting

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+// const defines
+//------------------------------------------------------------------------------
+
+//------------------------------------------------------------------------------
+// local types
+//------------------------------------------------------------------------------
+
+typedef struct
+{
+    BOOL                    fInitialized;      ///< True if socketwrapper instance shall be created.
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: Wrong comment? TRUE if socketwrapper instance is initialized, right?

File: blackchannel/POWERLINK/libs/psicommon/include/libpsicommon/global.h [hunk 7]
```diff
@@ -45,7 +46,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #ifdef PSI_BUILD_PCP
   #include <pcptarget/target.h>
-  #include <oplk/ami.h>
+  #include <libpsicommon/ami.h>
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/libs/psicommon/include/libpsicommon/global.h]: In both cases the libpsicommon/ami.h is included!?

File: blackchannel/POWERLINK/pcp/psi/event.c [hunk 8]
```diff
@@ -278,4 +269,45 @@ static tOplkError processErrorWarningEvent(tOplkApiEventType EventType_p,
...
+{
+    tObdAlConHdl*            pParam = pEventArg_p->userObdAccess.pUserObdAccHdl;
+    tOplkError               oplkret = kErrorOk;
+
+    UNUSED_PARAMETER(eventType_p);
+    UNUSED_PARAMETER(pUserArg_p);
+
+    switch(pParam->index)
+    {
+        #if(((PSI_MODULE_INTEGRATION) & (PSI_MODULE_SSDO)) != 0)
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/event.c]: @Alex: indenting here?

File: blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h [hunk 9]
```diff
@@ -60,11 +60,11 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 // function prototypes
 //------------------------------------------------------------------------------
 #if(((PSI_MODULE_INTEGRATION) & (PSI_MODULE_CC)) != 0)
-  extern tOplkError cc_obdAccessCb(tObdCbParam MEM* pParam_p);
+  extern tOplkError cc_obdAccessCb(tObdCbParam* pParam_p);
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h]: @Alex: indenting here?

File: blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h [hunk 10]
```diff
@@ -60,11 +60,11 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 #endif
 
 #if(((PSI_MODULE_INTEGRATION) & (PSI_MODULE_SSDO)) != 0)
-  extern tOplkError rssdo_obdAccessCb(tObdCbParam MEM* pParam_p);
+  extern tOplkError rssdo_obdAccessCb(tObdCbParam* pParam_p);
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h]: @Alex: indenting here?

File: blackchannel/POWERLINK/pcp/psi/logbook.c [hunk 11]
```diff
@@ -601,15 +603,17 @@ static tPsiStatus sendToDestTarget(tLogInstance pInstance_p,
...
+        /*case kErrorSdoUdpArpInProgress:
         {
             // ARP table is still not updated -> Retry to transmit the frame later!
             timeout_startTimer(pInstance_p->pArpTimeoutInst_m);
 
             pInstance_p->consTxState_m = kConsTxStateWaitForNextArpRetry;
 
             break;
-        }
+        }*/
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/logbook.c]: Commented code?!

File: blackchannel/POWERLINK/pcp/psi/main.c [hunk 12]
```diff
@@ -533,21 +590,27 @@ psi_syncCb() implements the event callback function called by event module
...
+    tPsiStatus          ret     = kPsiSuccessful;
+    tTimeInfo           time;
+    tNetTime*           pNetTime = NULL;
+    tOplkApiSocTimeInfo socTimeStamp;
+
+    ret = oplk_getSocTime(&socTimeStamp);
+    if(ret != kPsiSuccessful)
+    {
+      oplkret = kErrorInvalidOperation;
+      goto Exit;
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/main.c]: Correct indenting (4 spaces/indentation)

File: blackchannel/POWERLINK/pcp/psi/nios2.cmake [hunk 13]
```diff
@@ -76,7 +78,7 @@ SET(ALT_PSI_INCS
     ${ALT_PCP_BSP_DIR}
     ${ALT_PCP_BSP_DIR}/HAL/inc
     ${ALT_PCP_BSP_DIR}/drivers/inc
-    ${ARCH_SOURCE_DIR}/altera_nios2
+    ${ARCH_SOURCE_DIR}/altera-nios2
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/nios2.cmake]: Is that include still required!?

File: blackchannel/POWERLINK/pcp/psi/tssdo.c [hunk 14]
```diff
@@ -547,15 +547,15 @@ static tPsiStatus sendToDestTarget(tTssdoInstance pInstance_p,
...
+      /*  case kErrorSdoUdpArpInProgress:
         {
             // ARP table is still not updated -> Retry to transmit the frame later!
             timeout_startTimer(pInstance_p->pArpTimeoutInst_m);
 
             pInstance_p->consTxState_m = kConsTxStateWaitForNextArpRetry;
 
             break;
-        }
+        }*/
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/tssdo.c]: Commented code?

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c [hunk 15]
```diff
@@ -265,74 +238,186 @@ static void edrv2veth_changeAddress(UINT32 ipAddr_p, UINT32 subNetMask_p, UINT16
...
+//            P R I V A T E   F U N C T I O N S                               //
+//============================================================================//
+/// \name Private Functions
+/// \{
+
+//------------------------------------------------------------------------------
+/**
 \brief Free the packet buffer after transmission
 
 \param  pPacket       Pointer to the packet buffer for freeing
```
Comment (brunnera) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: doxygen parameter does not match with function signature

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+//------------------------------------------------------------------------------
+/**
+\brief  Close socket wrapper instance
+
+The function closes the given socket wrapper instance.
+
+\param  pSocketWrapper_p    Socket wrapper instance.
+\param  pSocketAddress_p    Socket address structure.
```
Comment (brunnera) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: there is no pSocketAddress_p parameter

File: blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h (hunk 9 above)
Comment (brunnera) [inline blackchannel/POWERLINK/pcp/psi/include/psi/obdict.h]: no indendation needed

File: blackchannel/POWERLINK/pcp/psi/status.c [hunk 17]
```diff
@@ -314,8 +315,8 @@ tPsiStatus status_process(tTimeInfo* pTime_p)
 //------------------------------------------------------------------------------
 void status_enableSyncInt(void)
 {
-    /* Enable the interrupt in the EPL status sync module*/
-    synctimer_enableExtSyncIrq(SYNC_INT_CYCLE_NUM, SYNC_INT_PULSE_WIDTH_NS);
+     /* Enable the interrupt in the EPL status sync module*/
```
Comment (brunnera) [inline blackchannel/POWERLINK/pcp/psi/status.c]: indentation

File: blackchannel/POWERLINK/pcp/psi/oplkcfg.h [hunk 18]
```diff
@@ -124,18 +140,18 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-#define DLLCAL_BUFFER_SIZE_TX_GEN_ASND              8192
-
-// Asynchronous transmit buffer for generic Ethernet frames in bytes
-#define DLLCAL_BUFFER_SIZE_TX_GEN_ETH               8192
-
 // Asynchronous transmit buffer for sync response frames in bytes
 #define CONFIG_DLLCAL_BUFFER_SIZE_TX_SYNC           4096
 
+// Asynchronous transmit buffer for virtual Ethernet frames in bytes
+#define CONFIG_DLLCAL_BUFFER_SIZE_TX_VETH           4096
```
Comment (zelenkaj) [inline blackchannel/POWERLINK/pcp/psi/oplkcfg.h]: If CONFIG_DLLCAL_QUEUE is set to DIRECT_QUEUE the CONFIG_DLLCAL_BUFFER_SIZE_* macros have no effect, since the dllcal-direct.c is used. Anyhow, what is the motivation of this change?

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c (hunk 6 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: Comment shall be changed to: ///< Flag to indicate whether socketwrapper instance is created

File: blackchannel/POWERLINK/libs/psicommon/include/libpsicommon/global.h (hunk 7 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/libs/psicommon/include/libpsicommon/global.h]: libpsicommon/ami.h shall be moved out from the #ifdef condition

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c (hunk 2 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: The TODO is placed to replace the members of structure **tEdrv2VethRxDesc** : length and aBuffer[IP_MTU] with the structure **ip_packet_typ** which has the members : unsigned long length and unsigned char data[2] from the file ip.h

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c (hunk 3 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: Okay.

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c (hunk 5 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: Okay.

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c (hunk 4 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: Okay.

File: blackchannel/POWERLINK/pcp/psi/main.c (hunk 12 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/main.c]: Okay.

File: blackchannel/POWERLINK/stacks/ip/edrv2veth.c (hunk 15 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/edrv2veth.c]: The function signature shall be changed.

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c (hunk 16 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: pSocketAddress_p shall be removed from the doxygen parameter.

File: blackchannel/POWERLINK/pcp/psi/oplkcfg.h (hunk 18 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/oplkcfg.h]: I will remove it as it is not used.

File: blackchannel/POWERLINK/pcp/psi/status.c (hunk 17 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/status.c]: Okay.

File: blackchannel/POWERLINK/stacks/ip/socketwrapper.c (hunk 1 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/stacks/ip/socketwrapper.c]: This comment is removed in the rebased changes.

This function is called when a new SDO connection is initialized to handle the Arp requests. During the pull request this support was not available in the base of the openPOWERLINK stack used hence this section was commented. 

Now the arp query related support are implemented in the openPOWERLINK V2.5 stack so this comment can be removed.

File: blackchannel/POWERLINK/pcp/psi/logbook.c (hunk 11 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/logbook.c]: During the pull request ARP request support was not available in the base of the openPOWERLINK stack used hence this section was commented.

Now the arp query related support are implemented in the openPOWERLINK V2.5 stack so this comment can be removed.

File: blackchannel/POWERLINK/pcp/psi/tssdo.c (hunk 14 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/tssdo.c]: During the pull request ARP request support was not available in the base of the openPOWERLINK stack used hence this section was commented.

Now the ARP query related support are implemented in the openPOWERLINK V2.5 stack so this comment can be removed.

File: blackchannel/POWERLINK/pcp/psi/nios2.cmake (hunk 13 above)
Comment (powerlink-team-kalycito) [inline blackchannel/POWERLINK/pcp/psi/nios2.cmake]: Not required. It shall be removed.
